proxy/certs/certs.go

```diff
@@ -215,9 +234,44 @@ func (s *RemoteCertSource) Remote(instance string) (cert *x509.Certificate, addr
 		logging.Errorf("%v", err)
 		logging.Errorf("WARNING: specifying the correct region in an instance string will become required in a future version!")
 	}
-	if len(data.IpAddresses) == 0 || data.IpAddresses[0].IpAddress == "" {
+
+	if len(data.IpAddresses) == 0 {
 		return nil, "", "", fmt.Errorf("no IP address found for %v", instance)
 	}
+
+	// Find the first matching IP address type between user input types and the types that the instance has
+	IPAddrInUse := ""
+	isFound := false
+	for _, eachIPAddrTypeByUser := range s.IPAddrTypes {
+		for _, eachIPAddrTypeOfInstance := range data.IpAddresses {
+			if strings.ToUpper(eachIPAddrTypeOfInstance.Type) == strings.ToUpper(eachIPAddrTypeByUser) {
+				IPAddrInUse = eachIPAddrTypeOfInstance.IpAddress
+				isFound = true
+				break
+			}
+		}
+
+		if isFound {
+			break
+		}
+	}
+
+	if IPAddrInUse == "" {
+		IPAddrTypesOfInstance := ""
+		for _, eachIPAddrTypeOfInstance := range data.IpAddresses {
+			IPAddrTypesOfInstance += "(TYPE: " + eachIPAddrTypeOfInstance.Type + " IP: " + eachIPAddrTypeOfInstance.IpAddress + ") "
+		}
+
+		IPAddrTypeOfUser := ""
+		for _, eachIPAddrOpt := range s.IPAddrTypes {
+			IPAddrTypeOfUser += eachIPAddrOpt + ","
+		}
+		IPAddrTypeOfUser = strings.TrimSuffix(IPAddrTypeOfUser, ",")
+
+		return nil, "", "", fmt.Errorf("User input IP address type (%v) does not match the instance %v, the instance's IP addresses are {%v}", IPAddrTypeOfUser, instance, IPAddrTypesOfInstance)
```

Comment:
Instead of accumulating the strings and joining with commas manually, you can just use Sprintf("%v") (which fmt.Errorf uses as well), for example:

https://play.golang.org/p/igOzq3tOvQ4

Reply:
Good idea! For data.IpAddresses, I still use for loop since it is a slice of pointers. Couldn't find a better way for now. Any comment is welcome!